homeassistant/components/tplink/switch.py

```diff
@@ -30,128 +30,79 @@ async def async_setup_entry(
-        self._async_update_attrs()
-
-    @async_refresh_after
-    async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the LED switch on."""
-        await self.device.set_led(True)
-
-    @async_refresh_after
-    async def async_turn_off(self, **kwargs: Any) -> None:
-        """Turn the LED switch off."""
-        await self.device.set_led(False)
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update the entity's attributes."""
-        is_on = self.device.led
-        self._attr_is_on = is_on
-        self._attr_icon = "mdi:led-on" if is_on else "mdi:led-off"
-
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
-        self._async_update_attrs()
-        super()._handle_coordinator_update()
-
-
-class SmartPlugSwitch(CoordinatedTPLinkEntity, SwitchEntity):
-    """Representation of a TPLink Smart Plug switch."""
-
-    _attr_name: str | None = None
+class Switch(CoordinatedTPLinkEntity, SwitchEntity):
+    """Representation of a feature-based TPLink sensor."""
 
     def __init__(
         self,
         device: SmartDevice,
         coordinator: TPLinkDataUpdateCoordinator,
-    ) -> None:
+        id_: str,
+        feature: Feature,
+        parent: SmartDevice = None,
+    ):
         """Initialize the switch."""
-        super().__init__(device, coordinator)
-        # For backwards compat with pyHS100
-        self._attr_unique_id = legacy_device_id(device)
+        super().__init__(device, coordinator, parent=parent)
+        self._device = device
+        self._feature = feature
+        self._attr_unique_id = f"{legacy_device_id(device)}_new_{id_}"
+        self._attr_entity_category = (
+            EntityCategory.CONFIG
+        )  # TODO: read from the feature
+        if feature.name == "State":  # Main switch of the device has no category.
+            self._attr_entity_category = None
+
+        self.entity_description = SwitchEntityDescription(
+            key=id_, name=feature.name, icon=feature.icon
+        )
         self._async_update_attrs()
 
     @async_refresh_after
     async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the switch on."""
-        await self.device.turn_on()
+        """Turn the LED switch on."""
+        await self._feature.set_value(True)
 
     @async_refresh_after
     async def async_turn_off(self, **kwargs: Any) -> None:
-        """Turn the switch off."""
-        await self.device.turn_off()
+        """Turn the LED switch off."""
+        await self._feature.set_value(False)
 
     @callback
     def _async_update_attrs(self) -> None:
         """Update the entity's attributes."""
-        self._attr_is_on = self.device.is_on
+        is_on = self._feature.value
+        self._attr_is_on = is_on
+
+        icon = self._feature.icon
+        # TODO: hacky way to support on/off icons. maybe this should be removed?
```

Comment:
I think we can use icon translations instead?

Reply:
Indeed, we could definitely do that. I suppose we cannot provide these progmatically, but they need to be defined in the JSON file?

Reply:
Right. They allow a default icon and one based on the current state

Example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/tessie/icons.json

Reply:
Yes, I think we don't want to do that until we find a workable solution for this. I tried to look into how we could check for the existence of translation, but couldn't find a quick solution. #103294 is the relevant PR, adding just here to have it written down somewhere.

Given it is just UI candy, we can do without it for the time being.